### PR TITLE
Refactor crawler function according to issue #16

### LINKED ISSUE #16: Modularize main script for scraping different years websites
The script `main.py` is used for scraping PyConTW 2016-2020 websites. Currently this script contains lots of global variables and shared functions which will increase the difficulty to keep flexible for scraping the official website in the future. It is  hard to maintain/develop. It will be great if we could make this script more structural like separating different year parsing detail into different handler or module, keep the same crawling processing in a base class, etc. Any ideas about this enhancement are welcome.

### PR DESCRIPTION
**Pull request by Darkborderman**
## Types of changes

- **Refactoring**

## Description
Refactor crawling function, fixes #16 

## Checklist:
- [x] Add test cases to all the changes you introduce
- [x] Run `inv style` locally to ensure all linter checks pass
- [x] Run `inv test` locally to ensure all test cases pass
- [x] Run `inv secure` locally to ensure no major vulnerability is introduced
- [x] Update the documentation if necessary

## Steps to Test This Pull Request
Currently don't have automated test, so use manual test
```bash
pipenv run python3 pycon_archive_past_website/pycon_archive_past_website.py -y 2020
```

## Expected behavior
Crawl 2020 pycon data

## Related Issue
#16 

## Additional context
None

### CODE REVIEW

File: pycon_archive_past_website/pycon_archive_past_website.py [hunk 1]
```diff
@@ -317,16 +79,18 @@ def main():
...
 def check_year(param, base_url):
     """Get Pycon Website According To the Year"""
-    global PYCON_YEAR
-    global BASE_URL  # TODO: [Refactor] to encapuslate all the global variable
-    BASE_URL = base_url
-    PYCON_YEAR = str(param.year)
-    if PYCON_YEAR >= "2016" and PYCON_YEAR <= "2020":
-        main()
+    year: str = str(param.year)
+    if year >= "2016" and year <= "2020":
```
Comment (allen91wu) [inline pycon_archive_past_website/pycon_archive_past_website.py:92]: I think we can remove the specific range of website year. For future, we just need to update the customize crawler in website folder.

**Review by allen91wu — CHANGES_REQUESTED** [review]
Thank you for your contribution. I just leave a comment. Please feel free to tell me any concern.

File: pycon_archive_past_website/common/dataio.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    Create folder from execute entry folder
+
+    Args:
+        path (str): relative path
+    """
+    path = urlparse(path).path
+    try:
+        # correct the path to directory path and be a local path
+        dir = "." + path[0 : path.rfind("/") + 1]
```
Comment (josix) [inline pycon_archive_past_website/common/dataio.py:18]: nit: using `pathlib.Path` here would get more expressive. I'm not quite sure if `"."` here is redundant, could you help me check this?
```suggestion
        dir = ("." / Path(path).parent).resolve()
```

**Review by josix — CHANGES_REQUESTED** [review]
Hello, @Darkborderman thanks for your hard work. I've made some comments on `common` and `pycon_archive_past_website.py` and I'll keep reviewing the other parts. Please feel free to have a discussion with me about the suggestions.

File: pycon_archive_past_website/pycon_archive_past_website.py (hunk 1 above)
Comment (Darkborderman) [reply pycon_archive_past_website/pycon_archive_past_website.py:92]: Thanks for the feedback, I'll update this part.

File: pycon_archive_past_website/pycon_archive_past_website.py [hunk 3]
```diff
@@ -1,260 +1,33 @@
...
-                mkdir(person["photo"])
-                writefile(person["photo"])
-
-
-def get_assets(path: Path):
-    """
-    Get all assets in the given path.
-    """
-    if not path.parts[0] == "/":
+def get_page(crawler: BaseCrawler, url: str):
```
[the remark this answers is not in the document]
Comment (Darkborderman) [reply pycon_archive_past_website/pycon_archive_past_website.py:12]: OK, I'll take some time to cope with this.

File: pycon_archive_past_website/pycon_archive_past_website.py [hunk 4]
```diff
@@ -269,43 +42,32 @@ def get_page(path):
...
+        if len(path_parts) >= 2 and path_parts[1] == crawler.year:
+            get_page(crawler, crawler_url)
+            get_page(crawler, crawler_url.replace("zh-hant", "en-us"))
 
+    # Get favicon on front page
+    soup = get_soup(f"{crawler.url}/{crawler.year}/zh-hant/")
     for link in soup.findAll("link", {"rel": "icon"}):
         if "href" in link.attrs:
-            get_assets(Path(urlparse(link["href"]).path))
+            get_asset(link["href"])
```
Comment (josix) [inline pycon_archive_past_website/pycon_archive_past_website.py:70]: I guess this part is also the execution belonging to the crawler that filters out the urls and decides which one should be scraped. I guess moving it into the `BaseCrawler` will look cleaner.

File: pycon_archive_past_website/websites/classes.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
+"""Crawler related classes"""
```
Comment (josix) [inline pycon_archive_past_website/websites/classes.py:1]: How about naming this folder and file as `crawlers/base.py`? `websites` and `classes` seem to be too general.

Comment (Darkborderman) [reply pycon_archive_past_website/websites/classes.py]: Updated.

File: pycon_archive_past_website/common/dataio.py (hunk 2 above)
Comment (Darkborderman) [reply pycon_archive_past_website/common/dataio.py:18]: Updated.

File: pycon_archive_past_website/common/scrap.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+"""Data crawling related functions"""
```
[the remark this answers is not in the document]
Comment (Darkborderman) [reply pycon_archive_past_website/common/scrap.py:1]: Updated.

File: pycon_archive_past_website/pycon_archive_past_website.py (hunk 1 above)
Comment (Darkborderman) [reply pycon_archive_past_website/pycon_archive_past_website.py:92]: Updated.

File: pycon_archive_past_website/pycon_archive_past_website.py (hunk 4 above)
Comment (Darkborderman) [reply pycon_archive_past_website/pycon_archive_past_website.py:70]: Moved L66~L70 to separate `BaseCrawler` class and rename it to `get_favicon` now.

File: pycon_archive_past_website/pycon_archive_past_website.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Darkborderman) [reply pycon_archive_past_website/pycon_archive_past_website.py:12]: Updated.

**Review by allen91wu — APPROVED** [review]
Looks good! Would you like to rebase some commit? Make commit history more readable.

**Review by josix — APPROVED** [review]
LGTM! Thank you @Darkborderman for your working on this. The update improves the readability of the codebase, and also made that maintainable. Thanks a lot.

**Comment by Darkborderman** [comment]
Thanks for the compliment!

**Pull request merged**